include/screen.h

```diff
@@ -1,589 +1,661 @@
+    // Returns the height of the current font
+
+    virtual int fontHeight()
+    {
+        int16_t x1, y1;
+        uint16_t w, h;
+        getTextBounds(String("W"), 0, 0, &x1, &y1, &w, &h);
+        return h;
+    }
+
+    // textHeight
+    //
+    // Returns the height of a string in screen pixels
+
+    virtual int textHeight(const String & str)
+    {
+        int16_t x1, y1;
+        uint16_t w, h;
+        getTextBounds(str, 0, 0, &x1, &y1, &w, &h);
+        return h;
+    }
+
+    // textWidth
+    //
+    // Returns the width of a string in screen pixels
+
+    virtual int textWidth(const String & str)
+    {
+        int16_t x1, y1;
+        uint16_t w, h;
+        getTextBounds(str, 0, 0, &x1, &y1, &w, &h);
+        return w;
+    }
+
+    // Render the current page into this screen.
+    void Update(bool bRedraw);
+
+    // Run the screen update loop (button handling + periodic redraw)
+    void RunUpdateLoop();
+
+    // Flip to the next page and handle effect-rotation pause/resume semantics.
+    // Safe to call from button handlers.
+    static void FlipToNextPage();
+
+  private:
+    // Page registry and page count helpers
+    static std::vector<std::unique_ptr<class Page>>& Pages();
+    static int ActivePageCount();
+
+    #if defined(TOGGLE_BUTTON_0)
+        Bounce2::Button _button0;
+    #endif
+    #if defined(TOGGLE_BUTTON_1)
+        Bounce2::Button _button1;
+    #endif
+};
+
+// Class specializations of the Screen class for various display types can implement hardware-specific versions of functions
+// like fillRect.  They also do any required initial setup of the display.
+
+#if USE_M5DISPLAY
+
+    // M5Screen
+    //
+    // Display code for the M5 based TFT displays on the M5 Stick, Stick C Plus, and Stack
+
+    #include <M5UnitLCD.h>
+
+    // M5Screen
+    //
+    // Screen class that supports the M5 devices
+
+    class M5Screen : public Screen
+    {
+      public:
+
+        M5Screen(int w, int h) : Screen(w, h)
+        {
+            M5.Lcd.fillScreen(GREEN16);
+        }
+
+        virtual void drawPixel(int16_t x, int16_t y, uint16_t color) override
+        {
+            M5.Lcd.drawPixel(x, y, color);
+        }
+
+        virtual void fillRect(int16_t x, int16_t y, int16_t w, int16_t h, uint16_t color) override
+        {
+            M5.Lcd.fillRect(x, y, w, h, color);
+        }
+
+        virtual void fillScreen(uint16_t color) override
+        {
+            M5.Lcd.fillScreen(color);
+        }
+    };
+#endif
+
+#if USE_TFTSPI
+    #include <TFT_eSPI.h>
+    #include <SPI.h>
+
+    // TFTScreen
+    //
+    // Screen class that works with the TFT_eSPI library for devices such as the S3-TFT-Feather
+    class TFTScreen : public Screen
+    {
+        TFT_eSPI tft;
+
+    public:
+
+        TFTScreen(int w, int h) : Screen(w, h)
+        {
+            tft.begin();
+
+            #ifdef TFT_BL
+                pinMode(TFT_BL, OUTPUT);                // REVIEW begin() might do this for us

```

Comment:
I wanted to remove this, but seem to recall a case of "black screen" that turned out to be backlighting, and am worried someone will get burned if we don't explicitly enable it - but I can't recall the module specifically.